File: unvisited_stones/unvisited_stones.rb
# Находит количество валунов, которые не посетит ни одна птица.
# 
# @param n [Integer] Количество птиц.
# @param stones [Integer] Общее количество валунов.
# @param birds [Array<Integer>] Массив с дистанциями перелёта каждой птицы.
# @return [Integer] Количество непосещённых валунов.

def unvisited_stones(n, stones, birds)
    visited = Array.new(stones + 1, false)
    birds.each do |bird|
      (bird..stones).step(bird) { |pos| visited[pos] = true }
    end
    visited[1..-1].count(false)
  end

File: unvisited_stones/UnvisitedStones.java
public class UnvisitedStones {
    /**
     * Находит количество валунов, которые не посетит ни одна птица.
     * 
     * @param N Количество птиц.
     * @param stones Общее количество валунов.
     * @param birds Массив с дистанциями перелёта каждой птицы.
     * @return Количество непосещённых валунов.
     */
    
    public int unvisitedStones(int N, int stones, int[] birds) {
        boolean[] visited = new boolean[stones + 1];
        for (int bird : birds) {
            for (int pos = bird; pos <= stones; pos += bird) {
                visited[pos] = true;
            }
        }
        int count = 0;
        for (int i = 1; i <= stones; i++) {
            if (!visited[i]) {
                count++;
            }
        }
        return count;
    }
}